crates/sui-framework/packages/bridge/sources/bridge.move

```diff
@@ -0,0 +1,344 @@
+        linked_table::push_back(&mut inner.bridge_records, key, BridgeRecord {
+            message,
+            verified_signatures: none(),
+            claimed: false,
+        });
+
+        // emit event
+        emit(TokenBridgeEvent {
+            message_type: message_types::token(),
+            seq_num: bridge_seq_num,
+            source_chain: inner.chain_id,
+            sender_address: address::to_bytes(tx_context::sender(ctx)),
+            target_chain,
+            target_address,
+            token_type: token_id,
+            amount: token_amount,
+        });
+    }
+
+    // Record bridge message approvals in Sui, called by the bridge client
+    // If already approved, return early instead of aborting.
+    public fun approve_bridge_message(
+        self: &mut Bridge,
+        message: BridgeMessage,
+        signatures: vector<vector<u8>>,
+    ) {
+        let inner = load_inner_mut(self);
+        let key = message::key(&message);
+
+        // retrieve pending message if source chain is Sui, the initial message must exist on chain.
+        if (message::message_type(&message) == message_types::token() && message::source_chain(&message) == inner.chain_id) {
+            let record = linked_table::borrow_mut(&mut inner.bridge_records, key);
+            assert!(record.message == message, EMalformedMessageError);
+            assert!(!record.claimed, EInvariantSuiInitializedTokenTransferShouldNotBeClaimed);
+
+            // If record already has verified signatures, it means the message has been approved.
+            // Then we exit early.
+            if (option::is_some(&record.verified_signatures)) {
+                emit(TokenTransferAlreadyApproved { message_key: key });
+                return
+            };
+            // verify signatures
+            committee::verify_signatures(&inner.committee, message, signatures);
+            // Store approval
+            record.verified_signatures = some(signatures)
+        } else {
+            // At this point, if this message is in bridge_records, we know it's already approved
+            // because we only add a message to bridge_records after verifying the signatures.
+            if (linked_table::contains(&inner.bridge_records, key)) {
+                emit(TokenTransferAlreadyApproved { message_key: key });
+                return
+            };
+            // verify signatures
+            committee::verify_signatures(&inner.committee, message, signatures);
+            // Store message and approval
+            linked_table::push_back(&mut inner.bridge_records, key, BridgeRecord {
+                message,
+                verified_signatures: some(signatures),
+                claimed: false
+            });
+        };
+        emit(TokenTransferApproved { message_key: key });
+    }
+
+    // Claim token from approved bridge message
+    // Returns Some(Coin) if coin can be claimed. If already claimed, return None
+    fun claim_token_internal<T>(
```

Comment:
Let's not forget to add a `locked` state check (relaying the comment from the other PR.